Fix `WrongScopeError` being thrown on Rails master:

Edouard-chin opened the pull request:
Fix `WrongScopeError` being thrown on Rails master:

- ### Problem

  Rails made a change in rails/rails@d4367eb72601f6e5bba3206443e9d141e9753af8
  and TestFixtures don't make use of `method_name` anymore, but simply
  `name`.

  `name` on RSpec raises a `WrongScopeError` when called within an
  example.

  This patch overrides `name` to return the example name instead.

jasnow commented:
Appears the fix triggered an existing rubocop error.

JonRowe commented on lib/rspec/rails/fixture_support.rb:
```diff
@@ -50,6 +50,10 @@ def self.proxy_method_warning_if_called_in_before_context_scope(method_name)
             end
           end
 
+          def name
```
If this is only needed on master, we should surround it with a version check, e.g:
```suggestion
          if Rails.version.to_f > 6
            def name
              @example
            end
          end
```

JonRowe commented on lib/rspec/rails/fixture_support.rb:
```diff
@@ -50,6 +50,10 @@ def self.proxy_method_warning_if_called_in_before_context_scope(method_name)
 
+          def name
+            @example
+          end
```
If this is only needed on master, we should surround it with a version check, e.g:
```suggestion
          if Rails.version.to_f > 6
            def name
              @example
            end
          end
```

JonRowe replied:
If this is only needed on master, we should surround it with a version check, e.g:
```suggestion
          if Rails.version.to_f > 6
            def name
              @example
            end
          end
```

Edouard-chin commented:
Thanks for the quick review. I made the suggested changes

JonRowe commented on lib/rspec/rails/fixture_support.rb:
```diff
@@ -50,6 +50,12 @@ def self.proxy_method_warning_if_called_in_before_context_scope(method_name)
             end
           end
 
+          if ::ActiveRecord.version >= Gem::Version.new('6.1.0')
```
Sorry you can't use `Gem::Version`, as we don't have a dependency on rubygems being available. A simple
```suggestion
          if Rails.version.to_f > 6.0
```
is preferred.

Edouard-chin replied:
Rubygem is part of the standard library since ruby 1.9. I can use `Rails.version.to_f` if you really want to but I don't see any real reason

JonRowe replied:
We don't load any part of the standard library unless 100% necessary as it causes environment poisoning (where someone might not have required the right file but we do, tests pass, but it fails outside of tests). There is also a speed benefit to not using rubygems which we ourselves utilise in our tests by disabling it (which gives us the speed benefit and tests we're not poisoning the env).

JonRowe replied:
e.g Its simpler to not use it so please don't :)

Edouard-chin replied:
Ok I'll use `Rails.version.to_f`, thanks for quick answer :)

Not trying to argue on such a small detail but `Rails.version` already loads rubygem 🤷‍♂ . But anyway having the syntax you proposed will make things consistent since that's what it's used in this library.

JonRowe commented on spec/rspec/rails/fixture_support_spec.rb:
```diff
@@ -13,5 +13,15 @@ module RSpec::Rails
     end
+
+    it "doesn't raise a WrongScopeError on Rails 6.1" do
+      allow(ActiveRecord).to receive(:version) { Gem::Version.new('6.1.0') }
```
Proposed change removes this necessity

JonRowe commented on spec/rspec/rails/fixture_support_spec.rb:
```diff
@@ -13,5 +13,15 @@ module RSpec::Rails
+        include FixtureSupport
+      end
+
+      expect { group.new.name }.to_not raise_error
```
This doesn't really prove the issue, what code triggers the name usage?

Edouard-chin replied:
Issue that triggered the problem is when enrolling fixture transaction https://github.com/rails/rails/blob/b4ab1f19d8bcc4639c7379b560f000f01753b5b0/activerecord/lib/active_record/test_fixtures.rb#L101

This proves the issue as I just want to make sure `name` can be called without raising an error.
If you feel more confortable I can write an integration test and actually load fixtures, but this test should remain as I want to make sure `name` can be called within a spec

JonRowe replied:
This should be demonstrated by loading a fixtures, its an implementation detail that `name` is available as we never call it.

JonRowe commented on spec/rspec/rails/fixture_support_spec.rb:
```diff
@@ -13,5 +13,15 @@ module RSpec::Rails
       end
     end
+
+    it "doesn't raise a WrongScopeError on Rails 6.1" do
```
If the spec only needs to run on a version of Rails it should be flagged as such as we have CI running against multiple Rails and Ruby versions.

JonRowe commented:
If you rebase this the rubocop build should be fixed

jasnow commented:
@Edouard-chin - Looking forward to this fixed version.

benoittgt commented on lib/rspec/rails/fixture_support.rb:
```diff
@@ -50,6 +50,12 @@ def self.proxy_method_warning_if_called_in_before_context_scope(method_name)
             end
           end
 
+          if ::Rails.version.to_f >= 6.1
```
I see the error on the CI. `Rails.version` is not good enough?

pirj replied:
It seems that in this context `Rails` will be resolved to `RSpec::Rails`.

Edouard-chin replied:
nop, it's just 2 tests stubbing the `Rails` module for I don't know what reason https://github.com/rspec/rspec-rails/pull/2215/files#diff-1279af43b24337b9b8a76fc3b3287f9aR199. Build should go green

jasnow commented:
Rails::VERSION::STRING.to_f

Edouard-chin commented:
Not too sure what's up with the jruby builds, otherwise it's 🍏

pirj commented:
https://github.com/flori/json/issues/392

benoittgt commented:
Thanks @pirj. I just reran the 2 builds, the issue on json seems to be fixed now.

benoittgt commented:
@Edouard-chin thanks a lot for the last commit. Do you think you can handle @JonRowe review? 

Thanks :)

Edouard-chin commented:
> Do you think you can handle @JonRowe review?

Sure happy to, but which comment did I not address ?  If you are referring to https://github.com/rspec/rspec-rails/pull/2215#discussion_r355616564 I added a test to load fixtures

benoittgt commented:
Sorry @Edouard-chin. I just noticed. @JonRowe are you ok with the last changes?

Cheers :)

JonRowe commented on spec/rspec/rails/fixture_support_spec.rb:
```diff
@@ -13,5 +13,27 @@ module RSpec::Rails
+      end
+
+      expect { group.new.name }.to_not raise_error
+    end
```
This test can be removed, the other is fine.

JonRowe commented on spec/rspec/rails/fixture_support_spec.rb:
```diff
@@ -13,5 +13,27 @@ module RSpec::Rails
+      expect { group.new.name }.to_not raise_error
+    end
+
+    it "setup_fixture successfuly" do
```
```suggestion
    it "will allow #setup_fixture to run successfully" do
```

JonRowe commented on spec/rspec/rails/fixture_support_spec.rb:
```diff
@@ -13,5 +13,27 @@ module RSpec::Rails
+    end
+
+    it "setup_fixture successfuly" do
+      skip if Rails.version.to_f < 6.1
```
```suggestion
```

Edouard-chin commented:
Thanks for the review. I addressed all the comment. I also got rid of the test you mentioned, though I disagree that calling `name` is an implementation detail.

`name` is the standard way in Rails to retrieve the name of a test/spec. By adding a test where we explicitly call `name` we make sure that RSpec-rails is compatible with how Rails deal with test names.

JonRowe commented:
> Thanks for the review. I addressed all the comment. I also got rid of the test you mentioned, though I disagree that calling name is an implementation detail.

> `name` is the standard way in Rails to retrieve the name of a test/spec. By adding a test where we explicitly call `name` we make sure that RSpec-rails is compatible with how Rails deal with test names. Anyway just my cents, it's your project :)

What we care about is that this can be used with Rails, testing `name` does not guarantee that, it just guarantees we respond to `name` as we never call `name` ourself it is pointless. What we needed (and you have provided) is an integration test showing that we work with Rails whatever Rails calls, which happens to be `name` in `master` but previously has not been.

jasnow commented:
Any update on this PR?

JonRowe commented:
It needs to be rebased to go green

Edouard-chin commented:
Rebased and CI is 🍏 . Thanks for your patience

JonRowe commented on spec/rspec/rails/fixture_support_spec.rb:
```diff
@@ -13,5 +13,16 @@ module RSpec::Rails
     end
+
+    it "will allow #setup_fixture to run successfully", if: Rails.version.to_f > 6.0 do
+
```
```suggestion

```
